src/swish/parallel.ss

```diff
@@ -0,0 +1,378 @@
+      (and (parallel:options is? x) x))
+    (define (maybe-single x)
+      (and (null? (cdr x)) (car x)))
+    (define (proc-error who x)
+      (errorf who "~s is not a procedure" x))
+    (cond
+     [(maybe-proc opt/f) =>
+      (lambda (f)
+        (cond
+         [(null? args)
+          (let ([arg1 f/arg1])
+            (check1 who arg1)
+            (proc1 who default-options f arg1))]
+         [else
+          (check* who f/arg1 args)
+          (proc* who default-options f (cons f/arg1 args))]))]
+     [(maybe-opt opt/f) =>
+      (lambda (opt)
+        (cond
+         [(maybe-proc f/arg1) =>
+          (lambda (f)
+            (cond
+             [(not (pair? args)) (errorf #f "incorrect number of arguments to ~s" who)]
+             [(maybe-single args) =>
+              (lambda (arg1)
+                (check1 who arg1)
+                (proc1 who opt f arg1))]
+             [else
+              (check* who (car args) (cdr args))
+              (proc* who opt f args)]))]
+         [else
+          (proc-error who f/arg1)]))]
+     [else
+      (proc-error who opt/f)]))
+
+  (define (check-uniform who first rest get-length check error-msg)
+    (check who first)
+    (let ([n (get-length first)])
+      (let lp ([rest rest])
+        (cond
+         [(null? rest) #t]
+         [(begin
+            (check who (car rest))
+            #f)]
+         [(fx= (get-length (car rest)) n) (lp (cdr rest))]
+         [else (errorf who error-msg first (car rest))]))))
+
+  (define (check-list who x)
+    (unless (list? x)
+      (errorf who "~s is not a proper list" x)))
+
+  (define (check-lists who first rest)
+    (check-uniform who first rest length check-list
+      "lists ~s and ~s differ in length"))
+
+  (define (check-vector who x)
+    (unless (vector? x)
+      (errorf who "~s is not a vector" x)))
+
+  (define (check-vectors who first rest)
+    (check-uniform who first rest vector-length check-vector
+      "lengths of input vectors ~s and ~s differ"))
+
+  (define ($parallel-map1 who opt f ls)
+    (scatter/gather! who opt f (list->vector ls) vector->list))
+
+  (define ($parallel-map* who opt f more)
+    (scatter/gather! who opt
+      (lambda (v*) (apply f v*))
+      (list->vector (apply map list more))
+      vector->list))
+
+  (define (parallel:map opt/f f/ls . args)
+    (parse-args 'parallel:map
+      $parallel-map1
+      $parallel-map*
+      opt/f f/ls args
+      check-list
+      check-lists))
+
+  (define parallel:execute
+    (case-lambda
+     [(thunks)
+      ($parallel-map1 'parallel:execute default-options thaw thunks)]
+     [(opt thunks)
+      ($parallel-map1 'parallel:execute (check-opt 'parallel:execute opt) thaw thunks)]))
+
+  (define ($parallel-vec1 who opt f v)
+    (scatter/gather! who opt f (vector-copy v) values))
+
+  (define ($parallel-vec* who opt f more)
+    (scatter/gather! who opt
+      (lambda (v*) (apply f (vector->list v*)))
+      (apply vector-map vector more)
```

Comment:
avoid the conversion of vector to list:
`(lambda (v*) (apply f v*))`
`(apply vector-map list more)`